aggregator/leader_flush_mgr.go

```diff
@@ -0,0 +1,341 @@
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	schema "github.com/m3db/m3aggregator/generated/proto/flush"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/clock"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/sync"
+
+	"github.com/uber-go/tally"
+)
+
+type leaderFlushManagerMetrics struct {
+	queueSize         tally.Gauge
+	flushTimesPersist instrument.MethodMetrics
+}
+
+func newLeaderFlushManagerMetrics(scope tally.Scope) leaderFlushManagerMetrics {
+	return leaderFlushManagerMetrics{
+		queueSize:         scope.Gauge("queue-size"),
+		flushTimesPersist: instrument.NewMethodMetrics(scope, "flush-times-persist", 1.0),
+	}
+}
+
+type leaderFlushManager struct {
+	sync.RWMutex
+
+	nowFn                  clock.NowFn
+	checkEvery             time.Duration
+	jitterEnabled          bool
+	workers                xsync.WorkerPool
+	flushTimesKeyFmt       string
+	flushTimesStore        kv.Store
+	flushTimesPersistEvery time.Duration
+	logger                 xlog.Logger
+	scope                  tally.Scope
+
+	rand                *rand.Rand
+	flushTimes          flushMetadataHeap
+	flushTimesKey       string
+	lastPersistAt       time.Time
+	flushedSincePersist bool
+	flushTask           *leaderFlushTask
+	metrics             leaderFlushManagerMetrics
+}
+
+func newLeaderFlushManager(opts FlushManagerOptions) roleBasedFlushManager {
+	nowFn := opts.ClockOptions().NowFn()
+	rand := rand.New(rand.NewSource(nowFn().UnixNano()))
+	instrumentOpts := opts.InstrumentOptions()
+	scope := instrumentOpts.MetricsScope()
+	mgr := &leaderFlushManager{
+		nowFn:                  nowFn,
+		checkEvery:             opts.CheckEvery(),
+		jitterEnabled:          opts.JitterEnabled(),
+		workers:                opts.WorkerPool(),
+		flushTimesKeyFmt:       opts.FlushTimesKeyFmt(),
+		flushTimesStore:        opts.FlushTimesStore(),
+		flushTimesPersistEvery: opts.FlushTimesPersistEvery(),
+		logger:                 instrumentOpts.Logger(),
+		scope:                  scope,
+		rand:                   rand,
+		lastPersistAt:          nowFn(),
+		metrics:                newLeaderFlushManagerMetrics(scope),
+	}
+	mgr.flushTask = &leaderFlushTask{mgr: mgr}
+	return mgr
+}
+
+func (mgr *leaderFlushManager) Open(shardSetID string) {
+	mgr.Lock()
+	mgr.flushTimesKey = fmt.Sprintf(mgr.flushTimesKeyFmt, shardSetID)
+	mgr.Unlock()
+}
+
+// Init initializes the leader flush manager by enqueuing all
+// the flushers in the buckets.
+func (mgr *leaderFlushManager) Init(buckets []*flushBucket) {
+	mgr.Lock()
+	mgr.flushTimes.Reset()
+	for bucketIdx, bucket := range buckets {
+		mgr.enqueueBucketWithLock(bucketIdx, bucket)
+	}
+	mgr.Unlock()
+}
+
+func (mgr *leaderFlushManager) Prepare(buckets []*flushBucket) (flushTask, time.Duration) {
+	var (
+		shouldFlush = false
+		duration    tally.Timer
+		flushers    []PeriodicFlusher
+		waitFor     = mgr.checkEvery
+	)
+	mgr.Lock()
```

Comment:
Any reason not to `defer mgr.Unlock()` here to avoid a bug being introduced in the future?  Seems we only save on a few extra CPU instructions at the bottom setting values by not using defer here.